Make interactive window and native editor take their fontSize and fontFamily from the settings in VS Code.

DavidKutu opened the pull request:
We get them in nativeEditor.tsx and in interactivePanel.tsx
to apply them as styles, but we also need to send them
as props to eventually assign them in the monaco editor.

For #7624

- [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [x] Title summarizes what is changing
- [x] Has a [news entry](https://github.com/Microsoft/vscode-python/tree/master/news) file (remember to thank yourself!)
- [ ] Appropriate comments and documentation strings in the code
- [ ] Has sufficient logging.
- [ ] Has telemetry for enhancements.
- [ ] Unit tests & system/integration tests are added/updated
- [ ] [Test plan](https://github.com/Microsoft/vscode-python/blob/master/.github/test_plan.md) is updated as appropriate
- [ ] [`package-lock.json`](https://github.com/Microsoft/vscode-python/blob/master/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed)
- [ ] The wiki is updated with any design decisions/details.

rchiodo commented on src/datascience-ui/history-react/interactivePanel.tsx:
```diff
@@ -67,6 +67,18 @@ export class InteractivePanel extends React.Component<IInteractivePanelProps, IM
+        const style: React.CSSProperties = {
+            fontSize: fontSize,
+            fontFamily: fontFamily
+        };
```
I don't think this is necessary. You can just use the var(--code-font-size) in the css file. Same in the nativeEditor.tsx

rchiodo commented on src/datascience-ui/history-react/interactivePanel.tsx:
```diff
@@ -67,6 +67,18 @@ export class InteractivePanel extends React.Component<IInteractivePanelProps, IM
+        let fontSize: number = 14;
+        let fontFamily: string = 'Consolas, \'Courier New\', monospace';
+        if (this.mainPanelRef.current) {
+            fontSize = parseInt(getComputedStyle(this.mainPanelRef.current).getPropertyValue('--code-font-size'), 10);
```
>getComputedStyle [](start = 32, length = 16)

I think it might be more performant to add this to the state object and return this in response to the getCssRequest. Right now it includes the theme and the css itself. You could add the font family and the font size to that as well. That way it would also change when the user changed their font size.

rchiodo replied:
You could change ICodeCssGenerator::generateThemeCss to generateThemeData and have it return the font data as well as the css.


---
In reply to: [330790206](https://github.com/microsoft/vscode-python/pull/7746#discussion_r330790206) [](ancestors = 330790206)

rchiodo replied:
Although I think the css gets updated too, so maybe the update would work with code as well.

---
In reply to: [330790747](https://github.com/microsoft/vscode-python/pull/7746#discussion_r330790747) [](ancestors = 330790747,330790206)

DavidKutu replied:
I'm trying it, but the value of rootCss in the state of nativeEditor (for example), is not changing unless I close the whole tab. I'll do the change but it will behave the same.

rchiodo replied:
We do update the isDark and monaco theme on update though

---
In reply to: [330801816](https://github.com/microsoft/vscode-python/pull/7746#discussion_r330801816) [](ancestors = 330801816)

rchiodo replied:
Oh the css is actually stuffed into the html we generate. I forgot

---
In reply to: [330802081](https://github.com/microsoft/vscode-python/pull/7746#discussion_r330802081) [](ancestors = 330802081,330801816)

rchiodo replied:
So it wouldn't update on change. However the request would. So if you want to have the font size update, we should use the idea I had.

---
In reply to: [330802213](https://github.com/microsoft/vscode-python/pull/7746#discussion_r330802213) [](ancestors = 330802213,330802081,330801816)

DavidKutu replied:
That response (IGetCssResponse) already has css in it. With that it should update right?

rchiodo replied:
I think so. As the style setter should use it below.

---
In reply to: [330812548](https://github.com/microsoft/vscode-python/pull/7746#discussion_r330812548) [](ancestors = 330812548)

rchiodo commented on src/datascience-ui/history-react/interactivePanel.tsx:
```diff
@@ -67,6 +67,18 @@ export class InteractivePanel extends React.Component<IInteractivePanelProps, IM
     public render() {
+        let fontSize: number = 14;
+        let fontFamily: string = 'Consolas, \'Courier New\', monospace';
+        if (this.state.rootCss) {
```
>rootCss [](start = 23, length = 7)

You should do all of this in the mainStateController and have a state for the fontSize/fontFamily when the getCssResponse is handled.

rchiodo replied:
That way you'd only do it when it needs an update

---
In reply to: [331161292](https://github.com/microsoft/vscode-python/pull/7746#discussion_r331161292) [](ancestors = 331161292)

DavidKutu replied:
Got it. We still need to close the tab to apply the changes tho.

rchiodo replied:
Ah that's because the font size and font family are only used in the constructor for the markdown. We might be able to update the font during componentDidUpdate

---
In reply to: [331173958](https://github.com/microsoft/vscode-python/pull/7746#discussion_r331173958) [](ancestors = 331173958)

rchiodo commented on src/datascience-ui/native-editor/nativeEditor.tsx:
```diff
@@ -88,6 +88,18 @@ export class NativeEditor extends React.Component<INativeEditorProps, IMainState
 
+        let fontSize: number = 14;
+        let fontFamily: string = 'Consolas, \'Courier New\', monospace';
+        if (this.state.rootCss) {
```
>rootCss [](start = 23, length = 7)

Same comment as the interactive. Do this parsing in the mainStateController when it receives the getCssResponse

rchiodo commented on src/datascience-ui/interactive-common/mainStateController.ts:
```diff
@@ -1163,8 +1173,25 @@ export class MainStateController implements IMessageHandler {
+            const fontFamilyIndex = response.css.indexOf('--code-font-family: ');
+
+            if (fontSizeIndex > -1) {
+                const fontSizeEndIndex = response.css.indexOf('px;', fontSizeIndex + 18);
```
>18 [](start = 85, length = 2)

I assume the 18 and 20 here are the length of --code-font-size and --code-font-family? Maybe make '--code-font-size' and '--code-font-family' const values and then take their length instead? It's a little hard to figure out where the 18 and 20 come from.

DavidKutu replied:
Good point, I'll get that on the next commit.

DavidKutu commented:
@DonJayamanne I ended up not using React Context, but I addressed your comment by passing an object as a prop instead of an individual value. Commit b7f514a makes that change.
